hubble/relay: Remove ReportOffline and refactor PeerManager

glrf opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.


This commit removes the `ReportOffline` method as the gRPC library is already able to notice broken connections and will try to reconnect on its own and uses exponential backoff to try again. While the `ReportOffline` feature might have speed up re-connection time in some cases, it potentially leaked goroutines, tried to reconnect far too often ignoring any backoff, and can result in unexpected interactions with the gRPC library.

The goroutine leak could be triggered by making more than one request every 5 seconds, when at least one peer is inaccessible. We spawned a goroutine for every request that will try and reconnect to the peer. However, we only closed a goroutine whenever a dial times out, which is every 5 second. So if the rate of request is higher than one request every 5 second, we slowly build up goroutines.

At the same time this change fixes two different issues:

* Whenever the relay reconnected to the peer service, it leaked all exiting connections, as it didn't close exiting connections when adding peers with the same name. We fix this by treating `update` and `add` notifications the same.
* There was a minor data race as `List` did not lock the peer when copying the connection (which might be updated by `connect` or `disconnect`)

kaworu commented on pkg/hubble/relay/observer/observer.go:
```diff
@@ -25,11 +25,9 @@ func isAvailable(conn poolTypes.ClientConn) bool {
-		return true
-	}
-	return false
+	state := conn.GetState()
+	return state != connectivity.TransientFailure &&
+		state != connectivity.Shutdown
```
If I understand correctly: when the state is `connectivity.Connecting` we return `false` before this patch and `true` after this patch. What is the rational behind this change?

glrf replied:
Because `connectivity.Connecting` IMO shouldn't be treated as unavailable. If we're actively trying to connect I think we should try to make the request, which will block until we either successfully reconnected or failed.

But I'm also oke with leaving it as it was.

kaworu commented on pkg/hubble/relay/observer/server.go:
```diff
@@ -318,6 +302,13 @@ func (s *Server) ServerStatus(ctx context.Context, req *observerpb.ServerStatusR
+				if len(unavailableNodes) < numUnavailableNodesReportMax {
+					unavailableNodes = append(unavailableNodes, p.Name)
+				}
+				mu.Unlock()
```
if `mu` is protecting `numConnectedNodes`, `numUnavailableNodes`, and `unavailableNodes`, then we should lock it when accessing them from the "main" goroutine in the `for` loop above right? (lines 286, 290-292, and 295).

glrf replied:
Good catch! Yes I we should

kaworu commented on pkg/hubble/relay/observer/server.go:
```diff
@@ -333,13 +324,6 @@ func (s *Server) ServerStatus(ctx context.Context, req *observerpb.ServerStatusR
 	}()
 	resp := &observerpb.ServerStatusResponse{
 		Version: build.RelayVersion.String(),
-		NumConnectedNodes: &wrapperspb.UInt32Value{
-			Value: numConnectedNodes,
-		},
-		NumUnavailableNodes: &wrapperspb.UInt32Value{
-			Value: numUnavailableNodes,
-		},
-		UnavailableNodes: unavailableNodes,
```
Why are theses lines moved?

glrf replied:
Because at that point the requesting goroutines are still updating these values. They're only ready once the `statueses` channel is closed. They are replaced by L342-L350

kaworu replied:
Ah right I overlooked that `g.Wait()` is called in another goroutine, thanks!

kaworu commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -709,6 +710,115 @@ func TestPeerManager(t *testing.T) {
+
+	assert.Eventually(t, func() bool {
+		peers := mgr.List()
+		return peers[0].Address.String() == "192.0.1.2:4244"
```
just to be less reliant on the current implementation
```suggestion
		return len(peers) == 1 && peers[0].Address.String() == "192.0.1.2:4244"
```

kaworu commented on pkg/hubble/relay/pool/manager_test.go:
```diff
@@ -709,6 +710,115 @@ func TestPeerManager(t *testing.T) {
+	assert.Eventually(t, func() bool {
+		return ccDialCount.Load() == 2
+	}, time.Minute, 10*time.Millisecond, "reconnect to  new address")
+	assert.Eventuallyf(t, func() bool {
```
nit
```suggestion
	assert.Eventually(t, func() bool {
```

kaworu reviewed (commented):
Thanks for the patch @glrf, some questions but overall LGTM. Would be great it the patch could be split into more commits to ease review :pray:

glrf committed:
hubble/relay: Remove ReportOffline

This commit removes the `ReportOffline` method as the gRPC library is
already able to notice broken connections and will try to reconnect on
its own and uses exponential backoff to try again. While the
`ReportOffline` feature might have speed up re-connection time in some
cases, it potentially leaked goroutines, tried to reconnect far too
often ignoring any backoff, and can result in unexpected interactions
with the gRPC library.

The goroutine leak could be triggered by making more than one request every
5 seconds, when at least one peer is inaccessible. We spawned a goroutine
for every request that will try and reconnect to the peer. However, we
only closed a goroutine whenever a dial times out, which is every 5
second. So if the rate of request is higher than one request every 5
second, we slowly build up goroutines.

Signed-off-by: Fabian Fischer <fabian.fischer@isovalent.com>

glrf committed:
hubble/relay: Fix connection leak when reconnecting to peer service

Whenever the relay reconnected to the peer service, it leaked all
exiting connections, as it didn't close exiting connections when
adding peers with the same name. We fix this by treating `update` and
`add` notifications the same.

Signed-off-by: Fabian Fischer <fabian.fischer@isovalent.com>

glrf committed:
hubble/relay: Fix data race in `List`

Signed-off-by: Fabian Fischer <fabian.fischer@isovalent.com>

glrf committed:
hubble/relay: Improve handling of offline nodes in `ServerStatus`

Before, `ServerStatus` would only report nodes as offline if they
already failed and where marked as offline. This could mean that the
`ServerStatus` response at the same time returned an error and reported
all nodes as online.

Signed-off-by: Fabian Fischer <fabian.fischer@isovalent.com>

gandro commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -236,20 +236,40 @@ func (m *PeerManager) List() []poolTypes.Peer {
 }
 
-func (m *PeerManager) add(hp *peerTypes.Peer) {
+func (m *PeerManager) put(hp *peerTypes.Peer) {
```
nit, informal (i.e. non-blocking): `put` is absolutely fine, but I wanted to point out that most other code in the Cilium codebase uses the name `upsert` for such functions

glrf replied:
I personally find `upsert` a bit fuggly so I usually use `put`, but if `upsert` is more common let's rename it to `upsert`.

gandro commented on pkg/hubble/relay/pool/manager.go:
```diff
@@ -236,20 +236,40 @@ func (m *PeerManager) List() []poolTypes.Peer {
 	}
 }
 
+func peerEqual(a, b peerTypes.Peer) bool {
```
I'm a bit worried about future changes to the peer type invalidating this function. Would it make sense to move this into `peerTypes`, and maybe even make it a method of the `Peer` struct, such that it is closer to the definition of its fields?

glrf replied:
Good point, adding an `Equal` method makes more sense.

gandro reviewed (approved):
Awesome work, some really nice catches! I've left some very minor feedback, feel free to ignore. Thanks!

kaworu commented on pkg/hubble/relay/observer/server.go:
```diff
@@ -276,18 +277,21 @@ func (s *Server) ServerStatus(ctx context.Context, req *observerpb.ServerStatusR
+			mu.Unlock()
 			continue
 		}
 		numConnectedNodes++
```
I think we need to hold `mu` when incrementing this one as well.

Overall, maybe we should only keep track of `numUnavailableNodes` and compute `numConnectedNodes` as `len(peers) - numUnavailableNodes` at the end. Both the math and locking would be a tad simpler, what do you think?

glrf replied:
Yes, you're absolutely right. And computing `numConnectedNodes` instead of counting it makes a lot of sense too. Thanks!

rolinh reviewed (approved):
More reliable, more tests, less code, less leak. Can we have more of that, please?
